stream-legal-entity/legal-entity-core/src/main/java/com/backbase/stream/LegalEntitySaga.java

```diff
@@ -228,7 +230,7 @@ private Mono<LegalEntityTask> processProducts(LegalEntityTask streamTask) {
 
         return Flux.fromIterable(legalEntity.getProductGroups())
             .map(actual -> createProductGroupTask(streamTask, actual))
-            .flatMap(productGroupStreamTask -> batchProductIngestionSaga.process(productGroupStreamTask)
+            .concatMap(productGroupStreamTask -> batchProductIngestionSaga.process(productGroupStreamTask)
```

Comment:
As this could impact performance, an interesting approach would be if we pre-process this collection first to group the items, where we could have groups of product groups, where the products groups would process in serial (concatMap), and the groups of groups in parallel (flatMap)

Reply:
Indeed, such optimisation might be performed. It turned out, that each product group model has reference for list of users. Taking that into account there could be tricky combinations like ({pg-1,  user-1}, {pg-2, user-2}, {pg-3, user-1, user-2} or {pg-1, user-1, user-2}, {pg-2, user-2, user-3}) that would require well tested algorithm. Maybe that can be done right after this change (since it will just increase performance)